admin/code/AddToCampaignHandler.php

```diff
@@ -0,0 +1,232 @@
+	protected $data;
+
+	/**
+	 * AddToCampaignHandler constructor.
+	 *
+	 * @param $editForm - The parent form that triggered this action
+	 * @param $data - The data submitted as part of that form
+	 */
+	public function __construct($editForm, $data) {
+		$this->editForm = $editForm;
+		$this->data = $data;
+	}
+
+	/**
+	 * Perform the action. Either returns a Form or performs the action, as per the class doc
+	 *
+	 * @return HTML|SS_HTTPResponse
+	 */
+	public function handle() {
+		$object = $this->getObject($this->data['ID'], $this->data['ClassName']);
+
+		if (empty($this->data['Campaign'])) {
+			return $this->Form($object)->forTemplate();
+		}
+		else {
+			return $this->addToCampaign($object, $this->data['Campaign']);
+		}
+	}
+
+	/**
+	 * Get what ChangeSets are available for an item to be added to by this user
+	 * @return ArrayList[ChangeSet]
+	 */
+	protected function getAvailableChangeSets() {
+		return ChangeSet::get()
+			->filter('State', ChangeSet::STATE_OPEN)
+			->filterByCallback(function($item) {
+				return ($item->canView());
+			});
+	}
+
+	/**
+	 * Safely get a DataObject from a client-supplied ID and ClassName, checking: argument
+	 * validity; existence; and canView permissions.
+	 *
+	 * @param $id int - The ID of the DataObject
+	 * @param $class - The Class of the DataObject
+	 * @return DataObject - The referenced DataObject
+	 * @throws SS_HTTPResponse_Exception
+	 */
+	protected function getObject($id, $class) {
+		$id = (int)$id;
+		$class = ClassInfo::class_name($class);
+
+		if (!$class || !is_subclass_of($class, 'DataObject') || !Object::has_extension($class, 'Versioned')) {
+			return $this->editForm->httpError(400, _t(
+				'AddToCampaign.ErrorGeneral',
+				'We apologise, but there was an error'
+			));
+		}
+
+		$object = $class::get()->byID($id);
+
+		if (!$object) return $this->editForm->httpError(404, _t(
+			'AddToCampaign.ErrorNotFound',
+			'That {Type} couldn\'t be found',
+			'',
+			['Type' => $class]
+		));
+
+		if (!$object->canView()) return $this->editForm->httpError(403, _t(
+				'AddToCampaign.ErrorItemPermissionDenied',
+				'It seems you don\'t have the necessary permissions to add {ObjectTitle} to a campaign',
+				'',
+				['ObjectTitle' => $object->Title]
+			)
+		);
+
+		return $object;
+	}
+
+	/**
+	 * Builds a Form that mirrors the parent editForm, but with an extra field to collect the ChangeSet ID
+	 *
+	 * @param $object DataObject - The object we're going to be adding to whichever ChangeSet is chosen
+	 * @return Form
+	 */
+	public function Form($object) {
+
+		$inChangeSets = array_unique(ChangeSetItem::get_for_object($object)->column('ChangeSetID'));
+		$changeSets = $this->getAvailableChangeSets()->map();
+
+		$campaignDropdown = DropdownField::create('Campaign', '', $changeSets);
+		$campaignDropdown->addExtraClass('noborder');
+		$campaignDropdown->setDisabledItems($inChangeSets);
+
+		$fields = new FieldList([$campaignDropdown]);
+
+		foreach ($this->data as $k => $v) {
```

Comment:
You can change this to a recursive walk (probably in another method) that knows how to reconstruct a set of Name[Field] hiddenfields for each nested array.